File: app/src/main/java/com/pr0gramm/app/ui/views/CommentsAdapter.java
package com.pr0gramm.app.ui.views;

import android.annotation.SuppressLint;
import android.content.Context;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.google.common.base.Optional;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableListMultimap;
import com.google.common.collect.ListMultimap;
import com.google.common.collect.Lists;
import com.google.common.collect.Maps;
import com.google.common.collect.Multimaps;
import com.google.common.collect.Ordering;
import com.pr0gramm.app.R;
import com.pr0gramm.app.api.pr0gramm.Api;
import com.pr0gramm.app.feed.Vote;
import com.pr0gramm.app.services.ThemeHelper;
import com.pr0gramm.app.util.AndroidUtility;

import org.joda.time.Hours;
import org.joda.time.Instant;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;

import butterknife.ButterKnife;
import gnu.trove.set.TLongSet;
import gnu.trove.set.hash.TLongHashSet;

import static com.google.common.base.Ascii.equalsIgnoreCase;
import static com.google.common.base.MoreObjects.firstNonNull;
import static java.util.Collections.emptyList;
import static java.util.Collections.emptyMap;
import static org.joda.time.Instant.now;

/**
 */
public class CommentsAdapter extends RecyclerView.Adapter<CommentsAdapter.CommentView> {
    private final boolean admin;
    private final String selfName;
    private ImmutableList<CommentEntry> comments;
    private Optional<String> op;
    private CommentActionListener commentActionListener;
    private long selectedCommentId;

    private final Instant scoreVisibleThreshold = now().minus(Hours.ONE.toStandardDuration());
    private TLongSet favedComments = new TLongHashSet();
    private boolean showFavCommentButton;

    public CommentsAdapter(boolean admin, String selfName) {
        this.admin = admin;
        this.selfName = selfName;

        setHasStableIds(true);
        set(emptyList(), emptyMap(), null);
    }

    public void set(Collection<Api.Comment> comments, Map<Long, Vote> votes, String op) {
        this.op = Optional.fromNullable(op);

        Map<Long, Api.Comment> byId = Maps.uniqueIndex(comments, Api.Comment::getId);
        this.comments = ImmutableList.copyOf(Lists.transform(sort(comments, op), comment -> {
            int depth = getCommentDepth(byId, comment);
            Vote baseVote = firstNonNull(votes.get(comment.getId()), Vote.NEUTRAL);
            return new CommentEntry(comment, baseVote, depth);
        }));

        notifyDataSetChanged();
    }

    public void setShowFavCommentButton(boolean showFavCommentButton) {
        this.showFavCommentButton = showFavCommentButton;
    }

    public void setSelectedCommentId(long id) {
        if (selectedCommentId != id) {
            selectedCommentId = id;
            notifyDataSetChanged();
        }
    }

    public void setFavedComments(TLongSet favedComments) {
        boolean hasChanged = !this.favedComments.equals(favedComments);
        if (hasChanged) {
            this.favedComments = favedComments;
            notifyDataSetChanged();
        }
    }

    @Override
    public CommentView onCreateViewHolder(ViewGroup parent, int viewType) {
        return new CommentView(LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.comment_layout, parent, false));
    }

    @Override
    public void onViewRecycled(CommentView holder) {
        super.onViewRecycled(holder);
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onBindViewHolder(CommentView view, int position) {
        CommentEntry entry = comments.get(position);
        Api.Comment comment = entry.comment;

        view.setCommentDepth(entry.depth);
        view.senderInfo.setSenderName(comment.getName(), comment.getMark());
        view.senderInfo.setOnSenderClickedListener(v -> doOnAuthorClicked(comment));

        AndroidUtility.linkify(view.comment, comment.getContent());

        // show the points
        if (admin
                || equalsIgnoreCase(comment.getName(), selfName)
                || comment.getCreated().isBefore(scoreVisibleThreshold)) {

            view.senderInfo.setPoints(getCommentScore(entry));
        } else {
            view.senderInfo.setPointsUnknown();
        }

        // and the date of the post
        view.senderInfo.setDate(comment.getCreated());

        // enable or disable the badge
        boolean badge = op.transform(op -> op.equalsIgnoreCase(comment.getName())).or(false);
        view.senderInfo.setBadgeOpVisible(badge);

        // and register a vote handler
        view.vote.setVote(entry.vote, true);
        view.vote.setOnVoteListener(v -> {
            boolean changed = doVote(entry, v);
            notifyItemChanged(position);
            return changed;
        });

        // set alpha for the sub views. sadly, setting alpha on view.itemView is not working
        view.comment.setAlpha(entry.vote == Vote.DOWN ? 0.5f : 1f);
        view.senderInfo.setAlpha(entry.vote == Vote.DOWN ? 0.5f : 1f);

        view.reply.setOnClickListener(v -> doAnswer(comment));

        view.copyCommentLink.setOnClickListener(v -> {
            if (commentActionListener != null) {
                commentActionListener.onCopyCommentLink(comment);
            }
        });

        Context context = view.itemView.getContext();
        if (comment.getId() == selectedCommentId) {
            int color = ContextCompat.getColor(context, R.color.selected_comment_background);
            view.itemView.setBackgroundColor(color);
        } else {
            AndroidUtility.setViewBackground(view.itemView, null);
        }

        if (view.kFav != null) {
            if (showFavCommentButton) {
                boolean isFavorite = favedComments.contains(comment.getId());

                if (isFavorite) {
                    int color = ContextCompat.getColor(context, ThemeHelper.primaryColor());
                    view.kFav.setColorFilter(color);
                    view.kFav.setImageResource(R.drawable.ic_favorite);
                } else {
                    int color = ContextCompat.getColor(context, R.color.grey_700);
                    view.kFav.setColorFilter(color);
                    view.kFav.setImageResource(R.drawable.ic_favorite_border);
                }

                view.kFav.setVisibility(View.VISIBLE);
                view.kFav.setOnClickListener(v ->
                        commentActionListener.onCommentMarkAsFavoriteClicked(comment, !isFavorite));
            } else {
                view.kFav.setVisibility(View.GONE);
            }
        }
    }

    private CommentScore getCommentScore(CommentEntry entry) {
        int score = entry.comment.getUp() - entry.comment.getDown();
        score += entry.vote.getVoteValue() - entry.baseVote.getVoteValue();
        return new CommentScore(score, entry.comment.getUp(), entry.comment.getDown());
    }

    private void doOnAuthorClicked(Api.Comment comment) {
        if (commentActionListener != null)
            commentActionListener.onCommentAuthorClicked(comment);
    }

    private void doAnswer(Api.Comment comment) {
        if (commentActionListener != null)
            commentActionListener.onAnswerClicked(comment);
    }

    private boolean doVote(CommentEntry entry, Vote vote) {
        if (commentActionListener == null)
            return false;

        boolean performVote = commentActionListener.onCommentVoteClicked(entry.comment, vote);
        if (performVote) {
            entry.vote = vote;
        }

        return performVote;
    }

    @Override
    public int getItemCount() {
        return comments.size();
    }

    @Override
    public long getItemId(int position) {
        return comments.get(position).comment.getId();
    }

    public void setCommentActionListener(CommentActionListener commentActionListener) {
        this.commentActionListener = commentActionListener;
    }

    public static class CommentView extends RecyclerView.ViewHolder {
        final TextView comment;
        final VoteView vote;
        final SenderInfoView senderInfo;
        final ImageView kFav, reply;
        final View copyCommentLink;


        public CommentView(View itemView) {
            super(itemView);

            // get the subviews
            comment = ButterKnife.findById(itemView, R.id.comment);
            vote = ButterKnife.findById(itemView, R.id.voting);
            senderInfo = ButterKnife.findById(itemView, R.id.sender_info);
            kFav = ButterKnife.findById(itemView, R.id.kfav);
            reply = ButterKnife.findById(itemView, R.id.reply);
            copyCommentLink = ButterKnife.findById(itemView, R.id.copy_comment_link);
        }

        public void setCommentDepth(int depth) {
            ((CommentSpacerView) itemView).setDepth(depth);
        }
    }

    public interface CommentActionListener {

        boolean onCommentVoteClicked(Api.Comment comment, Vote vote);

        void onAnswerClicked(Api.Comment comment);

        void onCommentAuthorClicked(Api.Comment comment);

        void onCommentMarkAsFavoriteClicked(Api.Comment comment, boolean markAsFavorite);

        void onCopyCommentLink(Api.Comment comment);
    }

    /**
     * "Flattens" a list of hierarchical comments to a sorted list of comments.
     *
     * @param comments The comments to sort
     */
    private List<Api.Comment> sort(Collection<Api.Comment> comments, String op) {
        ImmutableListMultimap<Long, Api.Comment> byParent =
                Multimaps.index(comments, Api.Comment::getParent);

        ArrayList<Api.Comment> result = new ArrayList<>();
        appendChildComments(result, byParent, 0, op);
        return result;
    }

    private void appendChildComments(List<Api.Comment> target,
                                     ListMultimap<Long, Api.Comment> byParent,
                                     long id, String op) {

        Ordering<Api.Comment> ordering = COMMENT_BY_CONFIDENCE;
        if (op != null) {
            ordering = Ordering.natural().reverse()
                    .onResultOf((Api.Comment c) -> op.equalsIgnoreCase(c.getName()))
                    .compound(ordering);
        }

        List<Api.Comment> children = ordering.sortedCopy(byParent.get(id));
        for (Api.Comment child : children) {
            target.add(child);
            appendChildComments(target, byParent, (int) child.getId(), op);
        }
    }

    private static int getCommentDepth(Map<Long, Api.Comment> byId, Api.Comment comment) {
        int depth = 0;
        while (comment != null) {
            depth++;
            comment = byId.get(comment.getParent());
        }

        return Math.min(8, depth);
    }

    private static final Ordering<Api.Comment> COMMENT_BY_CONFIDENCE =
            Ordering.natural().reverse().onResultOf(Api.Comment::getConfidence);

    private class CommentEntry {
        final Api.Comment comment;
        final Vote baseVote;
        final int depth;

        Vote vote;

        public CommentEntry(Api.Comment comment, Vote baseVote, int depth) {
            this.comment = comment;
            this.baseVote = baseVote;
            this.depth = depth;
            this.vote = baseVote;
        }
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/views/CommentsAdapter.kt
package com.pr0gramm.app.ui.views

import android.annotation.SuppressLint
import android.support.v4.content.ContextCompat
import android.support.v4.view.ViewCompat
import android.support.v7.recyclerview.extensions.ListAdapter
import android.support.v7.util.DiffUtil
import android.support.v7.widget.PopupMenu
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.github.salomonbrys.kodein.android.appKodein
import com.github.salomonbrys.kodein.instance
import com.google.common.base.Ascii.equalsIgnoreCase
import com.pr0gramm.app.R
import com.pr0gramm.app.api.pr0gramm.Api
import com.pr0gramm.app.orm.Vote
import com.pr0gramm.app.services.ThemeHelper
import com.pr0gramm.app.services.UserService
import com.pr0gramm.app.services.config.ConfigService
import com.pr0gramm.app.ui.AsyncListAdapter
import com.pr0gramm.app.ui.dialogs.ignoreError
import com.pr0gramm.app.util.*
import gnu.trove.map.TLongObjectMap
import gnu.trove.map.hash.TLongObjectHashMap
import org.joda.time.Hours
import org.joda.time.Instant.now
import rx.Observable
import kotlin.math.absoluteValue

/**
 */
class CommentsAdapter(
        private val admin: Boolean, private val selfName: String,
        private val actionListener: Listener) : AsyncListAdapter<CommentsAdapter.Entry, CommentView>(ItemCallback()) {

    private val scoreVisibleThreshold = now().minus(Hours.ONE.toStandardDuration())

    // the currently selected comment. Set to update comment
    var selectedCommentId by observeChangeEx(0L) { _, new ->
        state = state.copy(selectedCommentId = new)
    }

    private var state: State by observeChangeEx(State()) { old, new ->
        if (old != new) {
            updateVisibleComments()
        }
    }

    private var collapsed: Set<Long> by observeChange(setOf()) {
        state = state.copy(collapsed = collapsed)
    }

    init {
        setHasStableIds(true)
    }

    class ItemCallback : DiffUtil.ItemCallback<Entry>() {
        override fun areItemsTheSame(oldItem: Entry, newItem: Entry): Boolean {
            return oldItem.comment.id == newItem.comment.id
        }

        override fun areContentsTheSame(oldItem: Entry, newItem: Entry): Boolean {
            return oldItem == newItem
        }
    }

    fun updateVotes(votes: TLongObjectMap<Vote>) {
        val currentVotes = TLongObjectHashMap(votes)
        val baseVotes = state.baseVotes ?: TLongObjectHashMap(votes)

        state = state.copy(baseVotes = baseVotes, currentVotes = currentVotes)
    }

    fun updateComments(comments: Collection<Api.Comment>, op: String?) {
        state = state.copy(allComments = comments.toList(), op = op)
    }

    private fun updateVisibleComments() {
        val targetState = state

        // do not show
        if (targetState.allComments.isEmpty()) {
            submitList(listOf())
        }

        Observable.fromCallable { CommentTree(targetState).visibleComments }
                .ignoreError()
                .subscribeOnBackground().observeOnMain()
                .subscribe { visibleComments ->
                    // verify that we still got the right state
                    if (state === targetState) {
                        submitList(visibleComments)
                    }
                }
    }

    override fun getItemId(position: Int): Long {
        return getItem(position).comment.id
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CommentView {
        return CommentView(LayoutInflater
                .from(parent.context)
                .inflate(R.layout.comment_layout, parent, false))
    }

    @SuppressLint("SetTextI18n")
    override fun onBindViewHolder(holder: CommentView, position: Int) {
        val entry = getItem(position)
        val comment = entry.comment
        val context = holder.itemView.context

        // skip animations if this holders comment changed
        val holderChanged = holder.id.update(comment.id)

        holder.updateCommentDepth(entry.depth)
        holder.senderInfo.setSenderName(comment.name, comment.mark)
        holder.senderInfo.setOnSenderClickedListener {
            actionListener.onCommentAuthorClicked(comment)
        }

        AndroidUtility.linkifyClean(holder.comment, comment.content)

        // show the points
        if (admin
                || equalsIgnoreCase(comment.name, selfName)
                || comment.created.isBefore(scoreVisibleThreshold)) {

            holder.senderInfo.setPoints(entry.currentScore)
        } else {
            holder.senderInfo.setPointsUnknown()
        }

        // and the date of the post
        holder.senderInfo.setDate(comment.created)

        // enable or disable the badge
        holder.senderInfo.setBadgeOpVisible(state.op == comment.name)

        // and register a vote handler
        holder.vote.setVoteState(entry.vote, animate = !holderChanged)
        holder.vote.onVote = { vote -> doVote(entry, vote) }

        // set alpha for the sub views. sadly, setting alpha on view.itemView is not working
        holder.comment.alpha = if (entry.vote === Vote.DOWN) 0.5f else 1f
        holder.senderInfo.alpha = if (entry.vote === Vote.DOWN) 0.5f else 1f

        holder.reply.setOnClickListener {
            actionListener.onAnswerClicked(comment)
        }

        if (comment.id == selectedCommentId) {
            val color = ContextCompat.getColor(context, R.color.selected_comment_background)
            holder.itemView.setBackgroundColor(color)
        } else {
            ViewCompat.setBackground(holder.itemView, null)
        }

        holder.fav.let { fav ->
            val isFavorite = entry.vote == Vote.FAVORITE
            val newVote = if (isFavorite) Vote.UP else Vote.FAVORITE

            if (isFavorite) {
                val color = ContextCompat.getColor(context, ThemeHelper.accentColor)
                fav.setColorFilter(color)
                fav.setImageResource(R.drawable.ic_favorite)
            } else {
                val color = ContextCompat.getColor(context, R.color.grey_700)
                fav.setColorFilter(color)
                fav.setImageResource(R.drawable.ic_favorite_border)
            }

            fav.visible = true
            fav.setOnClickListener {                doVote(entry, newVote)            }
        }

        if (entry.hiddenCount != null) {
            holder.more.visible = false

            holder.expand.visible = true
            holder.expand.text = "+" + entry.hiddenCount
            holder.expand.setOnClickListener {
                collapsed -= comment.id
                notifyItemChanged(position)
            }

        } else {
            holder.expand.visible = false

            holder.more.visible = true
            holder.more.setOnClickListener { view ->
                showCommentMenu(view, entry)
            }
        }
    }

    private fun showCommentMenu(view: View, entry: Entry) {
        val userService = view.context.appKodein().instance<UserService>()

        val popup = PopupMenu(view.context, view)
        popup.inflate(R.menu.menu_comment)

        popup.menu.findItem(R.id.action_delete_comment)?.let { item ->
            item.isVisible = userService.userIsAdmin
        }

        popup.menu.findItem(R.id.action_collapse)?.let { item ->
            item.isVisible = entry.hasChildren && entry.comment.id !in collapsed
        }

        popup.setOnMenuItemClickListener { item -> onMenuItemClicked(item, entry) }
        popup.show()
    }

    private fun onMenuItemClicked(item: MenuItem, entry: Entry): Boolean {
        val l = actionListener

        when (item.itemId) {
            R.id.action_copy_link -> l.onCopyCommentLink(entry.comment)
            R.id.action_delete_comment -> l.onDeleteCommentClicked(entry.comment)
            R.id.action_collapse -> collapsed += entry.comment.id
        }

        return true
    }

    private fun doVote(entry: Entry, vote: Vote): Boolean {
        return actionListener.onCommentVoteClicked(entry.comment, vote)
    }

    data class State(
            val allComments: List<Api.Comment> = listOf(),
            val currentVotes: TLongObjectMap<Vote> = TLongObjectHashMap(),
            val baseVotes: TLongObjectMap<Vote>? = null,
            val collapsed: Set<Long> = setOf(),
            val op: String? = null,
            val selectedCommentId: Long = 0L)

    data class Entry(val comment: Api.Comment, val vote: Vote, val depth: Int,
                     val hasChildren: Boolean, val currentScore: CommentScore,
                     val hasOpBadge: Boolean, val hiddenCount: Int?)

    interface Listener {
        fun onCommentVoteClicked(comment: Api.Comment, vote: Vote): Boolean

        fun onAnswerClicked(comment: Api.Comment)

        fun onCommentAuthorClicked(comment: Api.Comment)

        fun onCopyCommentLink(comment: Api.Comment)

        fun onDeleteCommentClicked(comment: Api.Comment): Boolean
    }
}

class CommentView(itemView: View) : RecyclerView.ViewHolder(itemView) {
    val id = ValueHolder(0L)

    val vote: VoteView = itemView.find(R.id.voting)
    val reply: ImageView = itemView.find(R.id.action_reply)
    val comment: TextView = itemView.find(R.id.comment)
    val senderInfo: SenderInfoView = itemView.find(R.id.sender_info)

    val more: View = itemView.find(R.id.action_more)
    val fav: ImageView = itemView.find(R.id.action_kfav)
    val expand: TextView = itemView.find(R.id.action_expand)

    fun updateCommentDepth(depth: Int) {
        val spacerView = itemView as CommentSpacerView
        spacerView.depth = depth

        val maxLevels = ConfigService.get(itemView.context).commentsMaxLevels
        reply.visible = depth < maxLevels
    }
}

private class CommentTree(val state: CommentsAdapter.State) {
    private val byId = state.allComments.associateByTo(hashMapOf()) { it.id }
    private val byParent = state.allComments.groupByTo(hashMapOf()) { it.parent }

    private val depthCache = mutableMapOf<Long, Int>()

    /**
     * "Flattens" a list of hierarchical comments to a sorted list of comments.
     */
    private val linearizedComments: List<Api.Comment> = run {
        // bring op comments to top, then order by confidence.
        // actually we'll sort in reverse. We use this fact to optimize the
        // linearize step
        val ordering = compareBy<Api.Comment> { it.name == state.op }.thenBy { it.confidence }
        byParent.values.forEach { children -> children.sortWith(ordering) }


        mutableListOf<Api.Comment>().apply {
            val stack = byParent[0]?.toMutableList() ?: return@apply

            while (stack.isNotEmpty()) {
                // get next element
                val comment = stack.removeAt(stack.lastIndex)

                if (comment.id !in state.collapsed) {
                    // and add all children to stack if the element itself is not collapsed
                    byParent[comment.id]?.let { stack.addAll(it) }
                }

                // also add element to result
                add(comment)
            }
        }
    }

    val visibleComments: List<CommentsAdapter.Entry> = run {
        linearizedComments.map { comment ->
            val depth = depthOf(comment)
            val vote = currentVote(comment)
            val score = commentScore(comment, vote)

            val isCollapsed = comment.id in state.collapsed
            val hiddenCount = if (isCollapsed) countSubTreeComments(comment) else null

            val hasChildren = comment.id in byParent
            val hasOpBadge = state.op == comment.name

            CommentsAdapter.Entry(comment, vote, depth, hasChildren, score, hasOpBadge, hiddenCount)
        }
    }

    private fun currentVote(comment: Api.Comment): Vote {
        return state.currentVotes[comment.id] ?: Vote.NEUTRAL
    }

    private fun baseVote(comment: Api.Comment): Vote {
        return state.baseVotes?.get(comment.id) ?: Vote.NEUTRAL
    }

    private fun commentScore(comment: Api.Comment, currentVote: Vote): CommentScore {
        val delta = currentVote.voteValue - baseVote(comment).voteValue

        return CommentScore(
                comment.up + delta.coerceAtLeast(0),
                comment.down + delta.coerceAtMost(0).absoluteValue)
    }

    private fun countSubTreeComments(start: Api.Comment): Int {
        var count = 0

        val queue = mutableListOf(start)
        while (queue.isNotEmpty()) {
            val comment = queue.removeAt(queue.lastIndex)
            byParent[comment.id]?.let { children ->
                queue.addAll(children)
                count += children.size
            }
        }

        return count
    }

    private fun depthOf(comment: Api.Comment): Int {
        return depthCache.getOrPut(comment.id) {
            var current = comment
            var depth = 0

            while (true) {
                depth++

                // check if parent is already cached, then we'll take the cached value
                depthCache[current.parent]?.let { depthOfParent ->
                    return@getOrPut depth + depthOfParent
                }

                // it is not, lets move up the tree
                current = byId[current.parent] ?: break
            }

            return@getOrPut depth
        }
    }
}
